Rewrite to include additional features and be more language agnostic (#35)
RascalSoftware/RAT-Docs: source/tutorial/chapter2.rst

Defining a model is a case of populating the sections in the **Project** class, which we do using its methods.

In the following sections, we will look at each of these methods and see how they can be used to define our reflectivity analysis problem. (There is also a utility which will convert a RasCAL model into a **Project** for use in RAT.
This means that you can use the RasCAL GUI to set up your model if you like and then use the speed of RAT to conduct your analysis.)

In terms of the outputs, note that we have called the first output *problem*, but we don't need to do that. The first of the two outputs is another **Project**, but updated with the results of the calculation.
So, if we run a fit, the fitted parameters will be updated with the best fit values of our procedure. In giving the output the same name as the input, we're overwriting our input with best fit values, but if you don't
want to do this, you can give the inputs and outputs different names:

.. tab-set-code::
    .. code-block:: Matlab

>         [outputProblem, results] = RAT(problem, controls)

Reviewer: does this want a `;`?

there are always n+1 associated interfaces, and hence n+1 roughness parameters required. In RAT, the bulk interface roughness
is a protected parameter called ``"Substrate Roughness"`` and always exists, and this parameter controls where this roughness is placed in the layer stack. So, for two layers defined with thickness,
SLD and roughness as [d\ :sub:`1`, :math:`\rho_\mathrm{1}`, :math:`\sigma_\mathrm{1}`] and [d\ :sub:`2`, :math:`\rho_\mathrm{2}`, :math:`\sigma_\mathrm{1}`], then for the ``"substrate/liquid"`` geometry
the substrate roughness is placed as the first roughness the beam meets, and the layer roughness values refer to the interface after the particular layer.
But in the ``"air/substrate"`` case, the opposite is true, and the substrate roughness is the last roughness in the stack, with the layer roughness referring to the interface before each layer.

* **Model Type**: There are three ways of defining models in RAT:

    * **Standard Layers** - The model is defined in terms of parameters, which are distributed into layers, and subsequently grouped into contrasts. No external functions are needed.

>     * **Custom Layers** - Parameters are again defined and grouped into layers, but this time the layer definitions come from a user model script. This then gives complete flexibility of how layers are defined, so allowing models to be written in terms of area per molecule or material density, for example. This custom script controls translating these input parameters into a [d, :math:`\rho`, :math:`\sigma`] model. This is probably the most useful operating mode for RasCAL.

Reviewer: maybe just to explain, so that if a user isn't familiar with this notation they still get it?
```suggestion
    * **Custom Layers** - Parameters are again defined and grouped into layers, but this time the layer definitions come from a user model script. This then gives complete flexibility of how layers are defined, so allowing models to be written in terms of area per molecule or material density, for example. This custom script controls translating these input parameters into a [d, :math:`\rho`, :math:`\sigma`] (thickness, SLD, roughness) model. This is probably the most useful operating mode for RasCAL.
```
or something

For each entry in the table there are four fields:

* **Name**: The name you choose to give the datafile (for reference in the contrasts block)
* **Data**: An array containing the data itself (empty for 'Simulation').
* **Data Range**: The min / max range of the data you wish to include in the fit. You do not have to include all the data in the calculation of chi-squared. This range **must** lie **within** the range of any data added to the **Data** column.

> * **Simulation Range**: The total range of the simulation to be calculated. This can lie outside the range of the data to allow for extrapolation, but **must** be >= the range of any data added to the **Data** column.

Reviewer: ```suggestion
* **Simulation Range**: The total range of the simulation to be calculated. This must be equal to or larger than the range of any data added to the **Data** column.
```
or similar so it's plain English rather than "`>=`"

        problem.data.append(name='H-tail / D-head / ACMW', data=d13ACM)
        problem.data.append(name='D-tail / H-head / D2O', data=d70d2O)

We have everything we need to now build our contrasts. We have two contrasts in all, and we build them using name / value pairs for all the different parts of the contrasts (i.e. selecting which background and bulk phases etc we need using the names we have given them).

> To define the models for each contrast, we add the relevant layers as appropriate.

Reviewer: ```suggestion
To define the models for each contrast, we list the names of relevant layers as appropriate.
```
